### PR TITLE
Error space

### PR DESCRIPTION
**Pull request by MikeBishop**
Breaks the error space into different regions (#74) for the different components and starts defining error codes in each document.  (Note:  I didn't move anything into the loss-detection document, because I don't know what belongs there versus core transport.  Please help.)

Definitely need review and discussion here -- I may have deleted or moved error codes inappropriately.  Feel free to make changes to this branch.

### CODE REVIEW

**Comment by martinthomson** [comment]
Sorry Mike, but I'm taking the next week off.  I'll get back to this in the new year.

In the meantime, would you mind including a summary of what you think the design principles need to be.

**Comment by MikeBishop** [comment]
Hm.  Thought I had responded, but I don't see it here.

The basic principles I see are:

- The "owner" of a stream (TLS or app-layer) can reset that stream, or the connection, at any point for a reason it provides to QUIC.  QUIC will send this to the remote peer.
- QUIC can itself reset any stream or the entire connection for a reason that it surfaces to the app-layer and the remote peer.
- When a peer receives an error code in a RST_STREAM, GOAWAY, or CONNECTION_CLOSE, it needs:
    + A reasonable likelihood of figuring out what the error code means
    + A guarantee that it will *not* confuse the error code with some other error code.

To achieve this, I think we need to allocate a portion of the error space to QUIC, crypto, and the app-layer.  The remote knows the QUIC version, which implies the crypto protocol, and the application -- that means it can probably figure out what set of error codes to use for each region.  If extensions define new error codes, they would want to do so within the appropriate registries to avoid collisions and thereby confusion.

It's less clear whether congestion control / loss-detection should fall in the same bucket.  Since the remote doesn't know what congestion control algorithm I'm using, it doesn't have the same view into what set of errors would be appropriate.  I've currently allocated a chunk to it, but not moved any error codes.

Likewise, there might be value in defining an experimental range, either as a separate universe or as a subset of each range.  But then again, you could as easily have an experimental version of QUIC or the app-layer instead, which would have its own mappings and could define anything it wants.

**Comment by MikeBishop** [comment]
Jana's suggestion:  Split this PR in two; the purely-editorial of moving to the different documents, and the substantive of the divided space / renumbering.  Also, congestion control doesn't need its own space.

File: draft-ietf-quic-transport.md [hunk 1]
```diff
@@ -1639,245 +1639,190 @@ to get blocked.
...
 
-* 0x01: QUIC_INTERNAL_ERROR.  (Connection has reached an invalid state.)
+0x4000-0x7FFF:
+: Cryptographic error codes.  Defined by the crypto handshake protocol in use.
 
-* 0x02: QUIC_STREAM_DATA_AFTER_TERMINATION.  (There were data frames after the a
-  fin or reset.)
+0x8000-0xAFFF:
+: Loss-detection and congestion control error codes.  Defined by the
+  loss-detection and congestion control algorithm currently in use.
```
Comment (janaiyengar) [inline draft-ietf-quic-transport.md]: I don't think we need this fragment of error codes.

File: draft-ietf-quic-http.md [hunk 2]
```diff
@@ -690,55 +699,132 @@ following payload:
...
+: Not applicable, since QUIC handles stream management.  Would provoke a
+  QUIC_TOO_MANY_OPEN_STREAMS from the QUIC layer.
 
 CANCEL (0x8):
-: ?
+: HTTP_REQUEST_CANCELLED in {{http-error-codes}}.
 
 COMPRESSION_ERROR (0x9):
-: QUIC_DECOMPRESSION_FAILURE (not currently defined in core spec)
+: HTTP_HPACK_DECOMPRESSION_FAILEDin {{http-error-codes}}.
```
Comment (janaiyengar) [inline draft-ietf-quic-http.md]: Nit: "_FAILED in"

File: draft-ietf-quic-transport.md (hunk 1 above)
Comment (martinthomson) [inline draft-ietf-quic-transport.md]: Can you explain why?  Is it that congestion control is not supposed to generate errors?  Or do you expect to be able to take whatever error codes are needed from the core transport given that the two docs are more tightly coupled than the others?

**Comment by MikeBishop** [comment]
Well, there currently are no such error codes.  Plus, the peer doesn't know *which* congestion control algorithm you're using, so the point of having a variable lookup table is defeated.  I think I've been convinced that needs to just be folded in with Transport.

Jana made a good case for a host-local range, though, which MUST NOT appear on the wire, but can be used for logging errors and/or returning them in the API.

**Comment by MikeBishop** [comment]
Split into #130 and #131.

**Pull request closed without merging**